v2/tools/generator/internal/codegen/pipeline/create_types_for_backwards_compatibility.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+)
+
+const CreateTypesForBackwardCompatibilityID = "createTypesForBackwardCompatibility"
+
+// CreateTypesForBackwardCompatibility returns a pipeline stage that creates copies of types into other packages to
+// provide backward compatibility with previous releases of Azure Service Operator.
+// Backward compatibility versions are created for all versions to allow users of older versions of the operator to
+// easily upgrade.
+func CreateTypesForBackwardCompatibility(prefix string) *Stage {
+	stage := NewStage(
+		CreateTypesForBackwardCompatibilityID,
+		"Create clones of types for backward compatibility with prior ASO versions",
+		func(ctx context.Context, state *State) (*State, error) {
+			// Update the description of each to reflect purpose
+			withDescriptions, err := addCompatibilityComments(state.Definitions())
+			if err != nil {
+				return nil, errors.Wrap(err, "changing comments of types created for backward compatibility")
+			}
+
+			// Work out the new names for all our new types
+			renames := createBackwardCompatibilityRenameMap(state.Definitions(), prefix)
+
+			// Rename all the types
+			visitor := astmodel.NewRenamingVisitor(renames)
+
+			renamed, err := visitor.RenameAll(withDescriptions)
+			if err != nil {
+				return nil, errors.Wrap(err, "creating types for backward compatibility")
+			}
+
+			// Add the new types into our state
+			defs := state.Definitions()
+			defs.AddTypes(renamed)
+
+			return state.WithDefinitions(defs), nil
+		})
+
+	//
+	// Our pre-requisite support can't currently handle our needs because the earlier stage MUST be present, else the
+	// pipeline fails to run. These requirements are more of the "if the stage is present it must run before me" variety
```

Comment:
Is this in contrast to "the stage MUST EXIST and run before me"? It's weaker as it allows the stage to not exist?

Reply:
Yeah; if the stage is present at all, it must run first, but it's ok for it to be missing.